backends/ebpf/psa/README.md

```diff
@@ -164,6 +164,103 @@ a new element to the outer map (indexed by clone session or multicast group iden
+The PSA-eBPF compiler generates a BPF `LPM_TRIE` map instance for each P4 table instance. The hash map key as a concatenation of P4 match fields translated to eBPF representation.
+Moreover, the PSA-eBPF compiler shuffles the match fields and places the `lpm` field in the last position. Each `apply()` operation is translated into a lookup to the `LPM_TRIE` map.
+A control plane should populate the `LPM_TRIE` map with entries composed of a value and prefix. 
+
+### ternary
+
+There is no built-in BPF map for ternary (wildcard) matching. Hence, the PSA-eBPF compiler leverages the Tuple Space Search (TSS) algorithm for ternary matching (refer to the [research paper](https://dl.acm.org/doi/10.1145/316194.316216) to learn more about the TSS algorithm). 
+A `ternary` table is implemented using a combination of hash and array BPF maps that realizes the TSS algorithm. A P4 table is considered a `ternary` table if it contains at least one `ternary` field (exact and lpm fields are converted to ternary fields with an appropriate mask). 
+
+The PSA-eBPF compiler generates 2 BPF maps for each ternary table instance (+ the default action map):
+- the `<TBL-NAME>_prefixes` map is a BPF hash map that stores all unique ternary masks. The ternary masks are created based on the runtime table entries that are installed by a user.
+- the `<TBL-NAME>_tuples_map` map is a BPF array map of maps that stores all "tuples". A single tuple is a BPF hash map that stores all flow rules with the same ternary mask. 
+
+Note that the `psabpf-ctl table add` CLI command greatly simplifies the process of adding/removing flow rules to ternary tables. 
+
+For each `apply()` operation, the PSA-eBPF compiler generates the piece of code performing lookup to the above maps. The lookup code iterates over the `<TBL-NAME>_prefixes` map to 
+retrieve a ternary mask. Next, the lookup key (a concatenation of match keys) is masked with the obtained ternary mask and lookup to a corresponding tuple map is performed. 
+If a match is found, the best match with the highest priority is saved, and the algorithm continues to examine other tuples. If an entry with a higher priority is found in other tuples,
+the best match is overwritten. The algorithm exists when there is no more tuples to examine left.
+
+The snippet below shows the example of C code generated by the PSA-eBPF compiler for a lookup into a ternary table. The steps are explained below.
+
+```c
+struct ingress_tbl_ternary_1_key key = {};
+key.field0 = hdr->ipv4.dstAddr;
+key.field1 = hdr->ipv4.diffserv;
+struct ingress_tbl_ternary_1_value *value = NULL;
+struct ingress_tbl_ternary_1_key_mask head = {0};
+struct ingress_tbl_ternary_1_value_mask *val = BPF_MAP_LOOKUP_ELEM(ingress_tbl_ternary_1_prefixes, &head);
+if (val && val->has_next != 0) {
+    struct ingress_tbl_ternary_1_key_mask next = val->next_tuple_mask;
+    #pragma clang loop unroll(disable)
+    for (int i = 0; i < MAX_INGRESS_TBL_TERNARY_1_KEY_MASKS; i++) {  // (1)
```

Comment:
This works because the upper bound is a constant?

Reply:
Yes, exactly.